operator_cxx/contrib/quantization_int8-inl.h

```diff
@@ -0,0 +1,505 @@
+    CHECK_EQ(req.size(), 1U);
+    CHECK_EQ(req[Quantization_int8_enum::kOut], kWriteTo);
+
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 4, DType> data;
+    Tensor<xpu, 4, DType> out;
+    if (in_data[Quantization_int8_enum::kData].ndim() == 2) {
+      Shape<4> dshape = Shape4(in_data[Quantization_int8_enum::kData].shape_[0],
+                               in_data[Quantization_int8_enum::kData].shape_[1], 1, 1);
+      data = in_data[Quantization_int8_enum::kData].get_with_shape<xpu, 4, DType>(dshape, s);
+      out = out_data[Quantization_int8_enum::kOut].get_with_shape<xpu, 4, DType>(dshape, s);
+    } else {
+      data = in_data[Quantization_int8_enum::kData].get<xpu, 4, DType>(s);
+      out = out_data[Quantization_int8_enum::kOut].get<xpu, 4, DType>(s);
+    }
+    Tensor<xpu, 1, DType> aux = aux_states[Quantization_int8_enum::kMinmax].get<xpu, 1, DType>(s);
+
+    /*
+    for weight per channel, OIHW, the axis should be the O dimension
+    for all per channel, OIHW the axis should be the I dimension, currently this is not considered
+    transpose conv weight is also not considered
+    */
+    // print_host(data.dptr_, data.shape_.Size(), "in_data");
+    if (ctx.is_train > 0 && quant_countdown > 0) {
+      mshadow::Copy(out, data, s);
+      // Assign(out, req[Quantization_int8_enum::kOut], data);
+      quant_countdown = quant_countdown - 1;
+    }
+    else {
+      if (param_.quant_mode == std::string("minmax")) {
+        // if (param_.quant_mode == 0) {
+        // Assign(out, req[Quantization_int8_enum::kOut], data);
+        // mshadow::Copy(out, data, s);  
+        uint64_t WORKSPACE_LIMIT = 1024 * 1024; // allocate 1MB
+        Tensor<xpu, 1, uint8_t> workspace = ctx.requested[Quantization_int8_enum::kTempSpace]
+          .get_space_typed<xpu, 1, uint8_t>(Shape1(WORKSPACE_LIMIT), s);
+        uint64_t allocated_bytes = 0ULL;
+
+        mxnet::TShape src_shape, dst_shape;
+        const size_t temp_reduce_size = ConfigReduce<xpu, real_t>(
+            s, data.shape_, mxnet::TShape(1, 1), &src_shape, &dst_shape);
+        Tensor<xpu, 1, char> temp_reduce_space(reinterpret_cast<char*>(workspace.dptr_ + allocated_bytes), 
+                                       Shape1(temp_reduce_size), s);
+        allocated_bytes += temp_reduce_size;
+        const int dev_id = ctx.run_ctx.ctx.dev_id;
+        TBlob in_min_t(reinterpret_cast<real_t *>(workspace.dptr_ + allocated_bytes), Shape1(1), xpu::kDevMask,
+                      dev_id);
+        allocated_bytes += sizeof(real_t);
+        TBlob in_max_t(reinterpret_cast<real_t *>(workspace.dptr_ + allocated_bytes), Shape1(1), xpu::kDevMask,
+                      dev_id);
+        allocated_bytes += sizeof(real_t);
+
+        CHECK_LE(allocated_bytes, WORKSPACE_LIMIT) << " in forward";
+
+        // Tensor<xpu, 1, DType> min_val = in_min_t.get<xpu, 1, DType>(s);
+        Tensor<xpu, 1, DType> max_val = in_max_t.get<xpu, 1, DType>(s);
```

Comment:
Just use the primary type. No need to use a container for a single element.

Reply:
the tensor of `in_max_t` is for `mshadow::Copy` to assign  the value of `*in_max_t` to `aux`. i don't how to assign the value of `*in_max_t` to `aux` in easier way.

Reply:
You can directly index into aux's memory.

Reply:
when xpu=gpu, we can't assign the max_val to aux's memory in host code. Due to we use ema to update the aux, the temp val of `in_max_t` is needed.